File: ts-sdk/src/core/interfaces.ts
import { LogRecord } from './record';

export interface LogEnricher {
  enrich(record: LogRecord): LogRecord;
}

export interface LogHandler {
  handle(record: LogRecord): void;
  flush(): Promise<void>;
}

export interface LogFormatter {
  format(record: LogRecord): string;
}

export interface LogSampler {
  shouldSample(record: LogRecord): boolean;
}

export interface LoggerConfig {
  service: string;
  env: string;
  handlers?: LogHandler[];
  enrichers?: LogEnricher[];
  sampler?: LogSampler;
}

File: py-sdk/mylogger/core/interfaces.py
from abc import ABC, abstractmethod
from typing import Generic, TypeVar, List, Optional, Dict, Any, Protocol, runtime_checkable
from .record import LogRecord

T = TypeVar("T", bound=LogRecord)

@runtime_checkable
class LogEnricher(Protocol):
    def enrich(self, record: LogRecord) -> LogRecord:
        ...

@runtime_checkable
class LogHandler(Protocol):
    def handle(self, record: LogRecord) -> None:
        ...
    def flush(self, batch: List[LogRecord]) -> None:
        ...

@runtime_checkable
class LogFormatter(Protocol):
    def format(self, record: LogRecord) -> str:
        ...

@runtime_checkable
class LogSampler(Protocol):
    def should_sample(self, record: LogRecord) -> bool:
        ...
